src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1408,15 +1408,27 @@ public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
 
             public int hashCode() {
                 long t = utf8 ? 0 : Long.MAX_VALUE;
+                /*
+                 * The lastModifiedTime attribute is used to detect cases where
+                 * the same file is opened more than once and where it has been
+                 * modified in the mean-time
+                 */
                 t += attrs.lastModifiedTime().toMillis();
-                return ((int)(t ^ (t >>> 32))) + file.hashCode();
+                Object fk = attrs.fileKey();
+                return Long.hashCode(t) +
+                        (fk != null ? fk.hashCode() : file.hashCode());
             }
 
             public boolean equals(Object obj) {
                 if (obj instanceof Key key) {
                     if (key.utf8 != utf8) {
                         return false;
                     }
+                    /*
+                     * The lastModifiedTime attribute is used to detect cases where
+                     * the same file is opened more than once and where it has been
+                     * modified in the mean-time
+                     */
```

Comment:
The code change to the hashCode method is fine but I think the comments added to both hashCode + equals are still a bit confusing because they hint that the last modified time is used to detect if a zip file has been modified. What would you think about dropping these comments and instead put a class file comment on Key to say that it represents a key to a zip file. It key is based on the file key if available, or the canonical path if the file key is not available. The key is also based on the file's last modified time to allow for cases where a zip file is re-opened after it has been modified.